File: backend/src/backlog_manager_backend/csv/parse_csv.py
import csv
import io
import string
from decimal import Decimal

import msgspec
import structlog
from sqlalchemy.ext.asyncio import AsyncSession

from backlog_manager_backend.errors import (
    ConflictError,
    DatabaseError,
    NotFoundError,
    ValidationError,
)
from backlog_manager_backend.integrations.howlongtobeat import search_game_on_hltb
from backlog_manager_backend.repositories.backlog_entry_repo import create_backlog_entry
from backlog_manager_backend.schemas.backlog_entry import CreateBacklogEntryParams

logger = structlog.get_logger()

CSVRecord = dict[str, object]

_COLUMN_KEYS = list(string.ascii_uppercase)


class ColumnConfig(msgspec.Struct):
    title_column: str
    genre_column: str
    platform_column: str
    status_column: str


class MissingGame(msgspec.Struct):
    title: str
    genre: str
    platform: str
    status: str


class RecordError(msgspec.Struct):
    title: str
    error: str


class ImportResult(msgspec.Struct):
    success: int = 0
    failed: int = 0
    errors: list[RecordError] = []
    missing_games: list[MissingGame] = []


def parse_csv_content(file_content: str) -> list[CSVRecord]:
    reader = csv.reader(io.StringIO(file_content))
    records: list[CSVRecord] = []
    for row in reader:
        if not row or (len(row) == 1 and row[0] == ""):
            continue
        record: CSVRecord = {}
        for index, value in enumerate(row):
            if index < len(_COLUMN_KEYS):
                record[_COLUMN_KEYS[index]] = value
        records.append(record)
    return records


def _safe_string(value: object, default: str = "") -> str:
    if isinstance(value, str):
        return value
    if value is None:
        return default
    if isinstance(value, (int, float, bool)):
        return str(value)
    return default


_import_progress: dict[str, int] = {}
_import_cancel_flags: dict[str, bool] = {}


def get_import_progress(session_id: str) -> int:
    return _import_progress.get(session_id, 0)


def set_import_progress(session_id: str, processed: int) -> None:
    _import_progress[session_id] = processed


def clear_import_progress(session_id: str) -> None:
    _import_progress.pop(session_id, None)
    _import_cancel_flags.pop(session_id, None)


def set_cancel_flag(session_id: str, cancelled: bool) -> None:
    if cancelled:
        _import_cancel_flags[session_id] = True
    else:
        _import_cancel_flags.pop(session_id, None)


def get_cancel_flag(session_id: str) -> bool:
    return _import_cancel_flags.get(session_id, False)


def is_cancelled(session_id: str) -> bool:
    return get_cancel_flag(session_id)


async def import_backlog_entries_from_csv(
    session: AsyncSession,
    user_id: int,
    records: list[CSVRecord],
    config: ColumnConfig,
    session_id: str | None = None,
) -> ImportResult:
    result = ImportResult(errors=[], missing_games=[])
    processed_count = 0

    for record in records:
        if session_id and is_cancelled(session_id):
            logger.info(
                "Import cancelled",
                processed=processed_count + 1,
                total=len(records),
            )
            break

        title = _safe_string(record.get(config.title_column), "").strip()

        if not title:
            result.failed += 1
            result.errors.append(
                RecordError(
                    title="Unknown",
                    error=f"Title (column {config.title_column}) is required",
                )
            )
            processed_count += 1
            if session_id:
                set_import_progress(session_id, processed_count)
            continue

        search_results = await search_game_on_hltb(title)
        game_data = search_results[0] if search_results else None

        if game_data is None:
            result.missing_games.append(
                MissingGame(
                    title=title,
                    genre=_safe_string(record.get(config.genre_column), "Unknown"),
                    platform=_safe_string(record.get(config.platform_column), "Unknown"),
                    status=_safe_string(record.get(config.status_column), "Not Started"),
                )
            )
            processed_count += 1
            if session_id:
                set_import_progress(session_id, processed_count)
            continue

        try:
            await create_backlog_entry(
                session,
                CreateBacklogEntryParams(
                    user_id=user_id,
                    title=title,
                    genre=_safe_string(record.get(config.genre_column), "Unknown"),
                    platform=_safe_string(record.get(config.platform_column), "Unknown"),
                    status="Not Started",
                    owned=True,
                    interest=5,
                    image_link=game_data.image_url,
                    main_time=Decimal(str(game_data.main_story)),
                    main_plus_extra_time=Decimal(str(game_data.main_story_with_extras)),
                    completion_time=Decimal(str(game_data.completionist)),
                ),
            )
            result.success += 1
        except (NotFoundError, ConflictError, ValidationError, DatabaseError) as error:
            result.failed += 1
            result.errors.append(RecordError(title=title, error=str(error)))
        finally:
            processed_count += 1
            if session_id:
                set_import_progress(session_id, processed_count)

    return result

File: frontend/src/server/csv/parseCSV.ts
import { parse } from "csv-parse";
import { SearchGameOnHLTB } from "../integrations/howlongtobeat/howLongToBeat";
import type { Pool } from "pg";
import * as backlogEntryService from "~/server/services/backlogEntryService";

export type CSVRecord = Record<string, unknown>;

export interface ColumnConfig {
  titleColumn: string;
  genreColumn: string;
  platformColumn: string;
  statusColumn: string;
}

export interface BacklogEntryInput {
  title: string;
  genre: string;
  platform: string;
  status: string;
  owned: boolean;
  interest: number;
  releaseDate?: Date;
  imageLink?: string;
  mainTime?: number;
  mainPlusExtraTime?: number;
  completionTime?: number;
  reviewStars?: number;
  review?: string;
  note?: string;
}

export interface MissingGame {
  title: string;
  genre: string;
  platform: string;
  status: string;
}

export async function parseCSVContent(
  fileContent: string,
): Promise<CSVRecord[]> {
  return new Promise((resolve, reject) => {
    const records: unknown[] = [];

    const parser = parse(fileContent, {
      columns: false,
      skip_empty_lines: true,
    });

    parser.on("readable", function () {
      let record: unknown;
      while ((record = parser.read()) !== null) {
        records.push(record);
      }
    });

    parser.on("error", function (err) {
      reject(new Error(`CSV parsing error: ${err.message}`));
    });

    parser.on("end", function () {
      resolve(transformRecords(records));
    });
  });
}

function safeString(value: unknown, defaultValue = ""): string {
  if (typeof value === "string") return value;
  if (value === undefined || value === null) return defaultValue;
  if (typeof value === "number" || typeof value === "boolean")
    return String(value);
  return defaultValue;
}

function transformRecords(records: unknown[]): CSVRecord[] {
  const keys = [
    "A",
    "B",
    "C",
    "D",
    "E",
    "F",
    "G",
    "H",
    "I",
    "J",
    "K",
    "L",
    "M",
    "N",
    "O",
    "P",
    "Q",
    "R",
    "S",
    "T",
    "U",
    "V",
    "W",
    "X",
    "Y",
    "Z",
  ];

  return records.map((row) => {
    const transformedRow: CSVRecord = {};

    if (Array.isArray(row)) {
      row.forEach((value: unknown, index: number) => {
        const key = keys[index];
        if (key !== undefined) {
          transformedRow[key] = value;
        }
      });
    }

    return transformedRow;
  });
}

const importProgressMap = new Map<string, number>();
const importCancelFlagsMap = new Map<string, boolean>();

export function getImportProgress(sessionId: string): number {
  return importProgressMap.get(sessionId) ?? 0;
}

export function setImportProgress(sessionId: string, processed: number) {
  importProgressMap.set(sessionId, processed);
}

export function clearImportProgress(sessionId: string) {
  importProgressMap.delete(sessionId);
  importCancelFlagsMap.delete(sessionId);
}

export function setCancelFlag(sessionId: string, cancelled: boolean) {
  if (cancelled) {
    importCancelFlagsMap.set(sessionId, true);
  } else {
    importCancelFlagsMap.delete(sessionId);
  }
}

export function getCancelFlag(sessionId: string): boolean {
  return importCancelFlagsMap.get(sessionId) ?? false;
}

export function isCancelled(sessionId: string): boolean {
  return getCancelFlag(sessionId);
}

function emitProgress(sessionId: string, processed: number) {
  if (sessionId) {
    setImportProgress(sessionId, processed);
  }
}

export async function importBacklogEntriesFromCSV(
  pool: Pool,
  userId: number,
  records: CSVRecord[],
  config: ColumnConfig,
  sessionId?: string,
): Promise<{
  success: number;
  failed: number;
  errors: Array<{ title: string; error: string }>;
  missingGames: MissingGame[];
}> {
  const results = {
    success: 0,
    failed: 0,
    errors: [] as Array<{ title: string; error: string }>,
    missingGames: [] as MissingGame[],
  };

  let processedCount = 0;

  for (const record of records) {
    if (sessionId && isCancelled(sessionId)) {
      console.log(
        `Import cancelled at record ${processedCount + 1}/${records.length}`,
      );
      break;
    }

    try {
      const title = safeString(record[config.titleColumn], "").trim();

      if (!title) {
        results.failed++;
        results.errors.push({
          title: "Unknown",
          error: `Title (column ${config.titleColumn}) is required`,
        });
        processedCount++;
        if (sessionId) emitProgress(sessionId, processedCount);
        continue;
      }

      let gameData = null;
      let foundInHLTB = false;
      try {
        const searchResults = await SearchGameOnHLTB(title);
        if (searchResults && searchResults.length > 0) {
          gameData = searchResults[0];
          foundInHLTB = true;
        }
      } catch (searchError) {
        console.warn(
          `Could not find game "${title}" in howLongToBeat:`,
          searchError,
        );
      }

      if (!foundInHLTB) {
        results.missingGames.push({
          title,
          genre: safeString(record[config.genreColumn], "Unknown"),
          platform: safeString(record[config.platformColumn], "Unknown"),
          status: safeString(record[config.statusColumn], "Not Started"),
        });
        processedCount++;
        if (sessionId) emitProgress(sessionId, processedCount);
        continue;
      }

      //TODO Status is currently buggy because for now there are only enum states. Will be fixed in https://github.com/theoleuthardt/backlog-manager/issues/64#issue-3648683095
      const entryParams = {
        userId,
        title,
        genre: safeString(record[config.genreColumn], "Unknown"),
        platform: safeString(record[config.platformColumn], "Unknown"),
        status: "Not Started",
        owned: true,
        interest: 5,
        imageLink: gameData?.imageUrl,
        mainTime: gameData?.mainStory,
        mainPlusExtraTime: gameData?.mainStoryWithExtras,
        completionTime: gameData?.completionist,
      };

      try {
        console.log(`📝 Creating backlog entry:`, {
          title,
          genre: entryParams.genre,
          platform: entryParams.platform,
          status: entryParams.status,
        });
        await backlogEntryService.createBacklogEntry(pool, entryParams);
        results.success++;
        processedCount++;
        if (sessionId) emitProgress(sessionId, processedCount);
        console.log(`✓ Created backlog entry: ${title}`);
      } catch (createError) {
        console.error(
          `✗ Failed to create backlog entry for "${title}":`,
          createError,
        );
        throw createError;
      }
    } catch (error) {
      results.failed++;
      results.errors.push({
        title: safeString(record[config.titleColumn], "Unknown"),
        error: error instanceof Error ? error.message : "Unknown error",
      });
      processedCount++;
      if (sessionId) emitProgress(sessionId, processedCount);
    }
  }

  return results;
}
